### ISSUE TITLE
Crash with finalization

### ISSUE THREAD

**Issue by jburroni** [Cuis-Smalltalk/Machine-Learning#4]
If the image was saved with TF objects death but not yet finalized, then the image cannot be reopen

(don't know if it is a ML or Cuis'  issue) 

```
 0x7fff586cf750 I TF_Tensor>delete 0x10e2b15f0: a(n) TF_Tensor
    0x7fff586cf780 M TF_Tensor>finalize 0x10e2b15f0: a(n) TF_Tensor
    0x7fff586cf7b0 M [] in WeakRegistry>finalizeValues 0x10df2c5f8: a(n) WeakRegistry
    0x7fff586cf7f0 M OrderedCollection>do: 0x10d072ba0: a(n) OrderedCollection
    0x7fff586cf830 M WeakRegistry>finalizeValues 0x10df2c5f8: a(n) WeakRegistry
    0x7fff586cf860 M [] in WeakArray class>finalizationProcess 0x10d7bbbe8: a(n) WeakArray class
    0x7fff586cf8a8 M WeakArray(SequenceableCollection)>do: 0x10d6f4a90: a(n) WeakArray
    0x7fff586cf8e0 M [] in WeakArray class>finalizationProcess 0x10d7bbbe8: a(n) WeakArray class
    0x7fff586cf910 M BlockClosure>on:do: 0x10d071980: a(n) BlockClosure
    0x7fff586cf950 M BlockClosure>ifError: 0x10d071980: a(n) BlockClosure
```

**Comment by gerasdf** [comment]
Oh, it happened to me. I'll take a look, I'd say the dangling pointers
remain in memory, and when opened back they are finalizaed and are invalid
pointers.

I'd say this is a common issue, and I'd suggest doing a GC before saving.
It's strange that's not the case... If you have an easy way to reproduce it
could you test adding `Smalltalk garbageCollect` in the method before
saving the image?

**Comment by jburroni** [comment]
Actually, it worked correctly if I do GC before.

to reproduce it, open a Workspace and write:

```
nn _ NearestNeighbor new
```
and then save the image.

**Comment by jburroni** [comment]
Well, if you run the previously described example with `Smalltalk garbageCollect`, the image remains in a very _unstable_ state

**Comment by gerasdf** [comment]
Allright, got it:

To trigger it I did

   NearestNeighbor new.
   Smalltalk snapshot: true andQuit: true clearAllClassState: false

This saved the image, on entering again, a

   Smalltalk garbageCollect

will crash the system

To restore the image I used gdb, placed breakpoints in TF_GraphDelete and
TF_TensorDelete, and checked the arguments. They were all 0. So I skipped
the call (moving pc to the ret), and continued. The image was fine after
that.

Explanation:

On saving or restoring the image all ExternalAddresses are zeroed. On
reopening the image, any operation in this objects will crash the system,
they are external resources, this is "normal" behaviour. We may need to do
something else than just null the handles... not sure what. #become: them
all into a DeadExternalResource? not sure.

right now what I did is wrap all calls to librarie's delete function with a
"self isNull ifFalse: []"

This way finalization doesn't break anything...

cheers!
